### PR TITLE
toggle only items that really changed selection in multi mode

### LINKED ISSUE #75: In multi mode, already selected items fire deselect and then select when selection change occurs
The problem is due to the `selectMulti` logic:

    _selectMulti: function(values) {
      this._selection.clear();
      if (values) {
        for (var i = 0; i < values.length; i++) {
          this._selection.setItemSelected(this._valueToItem(values[i]), true);
        }
      }
    }

It first clear all selection, to reselect only the current selected values. The collateral effect is that if an item was already selected, it will fire a deselect event, and then fire a select event again (even though it was already selected)

### PR DESCRIPTION
**Pull request by reinert**
Fixes #75

### CODE REVIEW

File: iron-multi-selectable.html [hunk 1]
```diff
@@ -108,6 +115,12 @@
         this.splice('selectedValues',i,1);
       }
       this._selection.setItemSelected(this._valueToItem(value), unselected);
+    },
+
+    _valuesToItems: function(values) {
+      return (values == null) ? null : values.map(function(value) {
```
Comment (bicknellr) [inline iron-multi-selectable.html]: Any reason not to bind?

File: iron-multi-selectable.html [hunk 2]
```diff
@@ -81,11 +81,18 @@
...
-      this._selection.clear();
       if (values) {
-        for (var i = 0; i < values.length; i++) {
-          this._selection.setItemSelected(this._valueToItem(values[i]), true);
+        var selectedItems = this._valuesToItems(values);
+        // clear all but the current selected items
+        this._selection.clear(selectedItems);
+        // select only those not selected yet
+        for (var i = 0; i < selectedItems.length; i++) {
+          if (!this._selection.isSelected(selectedItems[i])) {
```
Comment (bicknellr) [inline iron-multi-selectable.html]: I think this conditional should be moved into [`IronSelection.setItemSelected`](https://github.com/PolymerElements/iron-selector/blob/a2a54b651faf48503d7196203fc6df7ae03bf81e/iron-selection.html#L73) (probably as `this._selection.indexOf(...) < 0`). For some reason, `IronSelection.setItemSelected`'s current implementation is inconsistent / lame: it's ok with trying to deselect an item that's not selected but it breaks when trying to select something that's already selected.

File: test/multi.html [hunk 3]
```diff
@@ -103,31 +103,106 @@
...
+
+        s.addEventListener('iron-select', function(e) {
+          selectEventCounters[items.indexOf(e.detail.item)]++;
+        });
+        s.addEventListener('iron-deselect', function(e) {
+          deselectEventCounters[items.indexOf(e.detail.item)]++;
+        });
+
+        // tap to select items 0 and 1 (both fire select)
+        items[0].dispatchEvent(new CustomEvent('tap', {bubbles: true}));
```
Comment (bicknellr) [inline test/multi.html]: We've got a thing for this: https://github.com/PolymerElements/iron-test-helpers/blob/66ad13aebf0dd4a58027175170af1f58a2f9fc36/mock-interactions.js#L174

It's already imported, so you can just do `MockInteractions.tap(items[0]);`.

**Comment by bicknellr** [comment]
Thanks for sending this PR; awesome tests! PTAL at the comments.

File: iron-multi-selectable.html (hunk 1 above)
Comment (reinert) [inline iron-multi-selectable.html]: I tried to follow the pattern [here](https://github.com/PolymerElements/iron-selector/blob/master/iron-selectable.html#L256)

**Comment by reinert** [comment]
Sorry for the delay. It's been a while and I needed to schedule a time to review it.

Thanks for appreciating the PR. I tried to follow your recommendations, except the first, which I stuck to the pattern already present at the code.

If you feel the bind is really beneficial to the code, I can change it too.

Thanks for your time.

**Comment by reinert** [comment]
Using MockInteractions on tests broke the patch. Some exception is occurring. Locally in my machine it is working fine. Do you know any reason why is that?

**Comment by reinert** [comment]
Checking master code, I saw sometimes using CustomEvent, other MockInteractions.

The difference is that MockInteractions is always used wrapping the source element with Polymer like `MockInteractions.tap(Polymer.dom(s).children[0])`. Is that trick mandatory?

[CustomEvent example](MockInteractions)
[MockInteractions example](MockInteractions)

**Comment by reinert** [comment]
Oh! After merging locally this patch with master I realized this failure is already present on master. So it's not an issue here.

File: iron-multi-selectable.html (hunk 1 above)
Comment (bicknellr) [inline iron-multi-selectable.html]: Oh, I'm not saying we should remove this function - mirroring IronSelectableBehavior's internals is totally a good thing here. I'm just curious about preferences between `thisArg` and binding; not important though.

Comment (bicknellr) [inline iron-multi-selectable.html]: What specific pattern are you referring to?

Comment (reinert) [inline iron-multi-selectable.html]: Now I got your point! ;)

Is it better now?

Comment (bicknellr) [inline iron-multi-selectable.html]: After looking around for a bit, it seems like there's still perf implications to using bind; I think we should switch this back to the `thisArg` approach. :/ Sorry for leading you astray there!

**Comment by bicknellr** [comment]
Left a comment. Also, #108 should fix the test, so once that's in you'll need to rebase to see it reflected here.

File: iron-multi-selectable.html (hunk 1 above)
Comment (reinert) [inline iron-multi-selectable.html]: No problem. I reset the commit.

**Comment by bicknellr** [comment]
#108 is in, could you rebase?

**Comment by reinert** [comment]
Done.

File: iron-multi-selectable.html [hunk 4]
```diff
@@ -86,11 +86,16 @@
...
 
     _selectMulti: function(values) {
-      this._selection.clear();
       if (values) {
-        for (var i = 0; i < values.length; i++) {
-          this._selection.setItemSelected(this._valueToItem(values[i]), true);
+        var selectedItems = this._valuesToItems(values);
+        // clear all but the current selected items
+        this._selection.clear(selectedItems);
+        // select only those not selected yet
```
Comment (bicknellr) [inline iron-multi-selectable.html]: This comment is no longer applicable.

Comment (bicknellr) [inline iron-multi-selectable.html]: Actually, nevermind. The code below seems to plainly counter what the comment says unless you look into the implementation of `setItemSelected`; but, the point of a comment is to make things that aren't obvious from the implementation clear, right?

**Comment by bicknellr** [comment]
LGTM, thanks again!

**Comment by reinert** [comment]
Thanks for reviewing the PR. You guys are doing an amazing job with the polymer project. My codes were never so beautiful. I'm definitely in love with it.

**Comment by bicknellr** [comment]
:)

**Pull request merged**